// yx-drawer/router/router.js
import routers from './index.js'

export function navigateTo(name, options = {}) {
	const url = getRouterUrl(name, options)
	uni.navigateTo({
		url: url,
		success: res => {},
		fail: () => {},
		complete: () => {}
	});
}

export function redirectTo(name, options = {}) {
	const url = getRouterUrl(name, options)
	uni.redirectTo({
		url: url,
		success: res => {},
		fail: () => {},
		complete: () => {}
	});
}

export function relaunch(name, options = {}) {
	const url = getRouterUrl(name, options)
	uni.reLaunch({
		url: url,
		success: res => {},
		fail: () => {},
		complete: () => {}
	});
}

export function switchTab(name, options = {}) {
	const url = getRouterUrl(name, options)
	uni.switchTab({
		url: url,
		success: res => {},
		fail: () => {},
		complete: () => {}
	});
}

// 将对象转为query参数的形式
// name 是跳转的路由名字，options是对象
export const getRouterUrl = (name, options) => {
	const _url = routers[name]
	const _keys = Object.keys(options || {})
	const _arr = []
	for (let i in _keys) {
		const _item = _keys[i]
		_arr.push(`${_item}=${options[_item]}`)
	}
	if (_arr && _arr.length > 0) {
		_url += `?` + _arr.join('&')
	}
	return _url
}

export default {
	navigateTo,
	redirectTo,
	relaunch,
	switchTab
}
